src/ledger/LedgerStateCache.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright 2025 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerStateCache.h"
+#include "ledger/LedgerHashUtils.h"
+#include "util/GlobalChecks.h"
+#include "util/Logging.h"
+#include "util/types.h"
+#include "xdrpp/printer.h"
+
+namespace stellar
+{
+size_t
+LedgerEntryHash::operator()(LedgerEntry const& entry) const
+{
+    // Extract the key from the entry and return the key's hash.
+    const LedgerKey& key = LedgerEntryKey(entry);
+    return std::hash<stellar::LedgerKey>()(key);
+}
+
+std::optional<LedgerEntry>
+LedgerStateCache::getEntry(LedgerKey const& key) const
+{
+    if (!supportedKeyType(key.type()))
+    {
+        return std::nullopt;
```

Comment:
This should be `releaseAssertOrThrow`. Otherwise, this return value is ambiguous, did we just look up a wrong type, or did we look up the correct type, but the key does not exist?